File: util/excel/src/main/java/cn/bestwu/simpleframework/util/excel/ExcelImport.java
package cn.bestwu.simpleframework.util.excel;

import cn.bestwu.simpleframework.util.excel.ExcelImportException.CellError;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import javax.validation.Validation;
import javax.validation.Validator;
import javax.validation.groups.Default;
import org.dhatim.fastexcel.reader.ReadableWorkbook;
import org.dhatim.fastexcel.reader.Row;
import org.dhatim.fastexcel.reader.Sheet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.multipart.MultipartFile;

/**
 * 导入Excel文件
 */
public class ExcelImport {

  private static final Logger log = LoggerFactory.getLogger(ExcelImport.class);
  private static final Validator validator = Validation.buildDefaultValidatorFactory()
      .getValidator();
  /**
   * 工作表对象
   */
  private final ReadableWorkbook workbook;
  /**
   * 验证 groups
   */
  private Class<?>[] validateGroups = new Class[]{Default.class};
  /**
   * 当前行号
   */
  private int rowNum;

  /**
   * 构造函数
   *
   * @param fileName 导入文件
   * @throws IOException IOException
   */
  public ExcelImport(String fileName)
      throws IOException {
    this(new File(fileName));
  }

  /**
   * 构造函数
   *
   * @param file 导入文件对象
   * @throws IOException IOException
   */
  public ExcelImport(File file)
      throws IOException {
    this(new FileInputStream(file));
  }

  /**
   * 构造函数
   *
   * @param multipartFile 导入文件对象
   * @throws IOException IOException
   */
  public ExcelImport(MultipartFile multipartFile)
      throws IOException {
    this(multipartFile.getInputStream());
  }

  /**
   * 构造函数
   *
   * @param is is
   * @throws IOException IOException
   */
  public ExcelImport(InputStream is)
      throws IOException {
    workbook = new ReadableWorkbook(is);
    log.debug("Initialize success.");
  }

  public ReadableWorkbook getWorkbook() {
    return workbook;
  }

  public int getRowNum() {
    return rowNum;
  }

  /**
   * 获取导入数据列表
   *
   * @param <F>         F
   * @param <E>         E
   * @param excelFields excelFields
   * @return List
   * @throws IOException            IOException
   * @throws IllegalAccessException IllegalAccessException
   * @throws ExcelImportException   ExcelImportException
   * @throws InstantiationException InstantiationException
   */
  public <F, E> List<E> getData(ExcelField<F, ?>[] excelFields)
      throws IOException, IllegalAccessException, InstantiationException, ExcelImportException {
    return getData(excelFields[0].entityType, excelFields);
  }


  /**
   * 获取导入数据列表
   *
   * @param converter   F 转换为E
   * @param <F>         F
   * @param <E>         E
   * @param excelFields excelFields
   * @return List
   * @throws IOException            IOException
   * @throws IllegalAccessException IllegalAccessException
   * @throws ExcelImportException   ExcelImportException
   * @throws InstantiationException InstantiationException
   */
  public <F, E> List<E> getData(ExcelField<F, ?>[] excelFields, ExcelConverter<F, E> converter)
      throws IOException, IllegalAccessException, InstantiationException, ExcelImportException {
    return getData(0, 0, excelFields[0].entityType, excelFields, converter);
  }

  /**
   * 获取导入数据列表
   *
   * @param <F>         F
   * @param <E>         E
   * @param excelFields excelFields
   * @param cls         实体类型
   * @return List
   * @throws IOException            IOException
   * @throws IllegalAccessException IllegalAccessException
   * @throws ExcelImportException   ExcelImportException
   * @throws InstantiationException InstantiationException
   */
  @SuppressWarnings("unchecked")
  public <F, E> List<E> getData(Class<F> cls, ExcelField<F, ?>[] excelFields)
      throws IOException, IllegalAccessException, InstantiationException, ExcelImportException {
    return getData(cls, excelFields, (o) -> (E) o);
  }


  /**
   * 获取导入数据列表
   *
   * @param converter   F 转换为E
   * @param <F>         F
   * @param <E>         E
   * @param excelFields excelFields
   * @param cls         实体类型
   * @return List
   * @throws IOException            IOException
   * @throws IllegalAccessException IllegalAccessException
   * @throws ExcelImportException   ExcelImportException
   * @throws InstantiationException InstantiationException
   */
  public <F, E> List<E> getData(Class<F> cls, ExcelField<F, ?>[] excelFields,
      ExcelConverter<F, E> converter)
      throws IOException, IllegalAccessException, InstantiationException, ExcelImportException {
    return getData(0, 0, cls, excelFields, converter);
  }


  /**
   * 获取导入数据列表
   *
   * @param headerNum   标题行号，数据行号=标题行号+1
   * @param sheetIndex  工作表编号
   * @param cls         实体类型
   * @param excelFields excelFields
   * @param converter   F 转换为E
   * @param <F>         F
   * @param <E>         E
   * @return List
   * @throws IOException            IOException
   * @throws IllegalAccessException IllegalAccessException
   * @throws ExcelImportException   ExcelImportException
   * @throws InstantiationException InstantiationException
   */
  public <F, E> List<E> getData(int sheetIndex, int headerNum, Class<F> cls,
      ExcelField<F, ?>[] excelFields,
      ExcelConverter<F, E> converter)
      throws IOException, IllegalAccessException, ExcelImportException, InstantiationException {
    Sheet sheet = workbook.getSheet(sheetIndex).orElse(null);
    if (sheet == null) {
      throw new RuntimeException("文档中未找到相应工作表!");
    }
    List<E> dataList = new ArrayList<>();
    for (Row row : sheet.openStream().filter(r -> r.getRowNum() - 1 > headerNum)
        .collect(Collectors.toList())) {
      if (row != null) {
        E e = readRow(cls, excelFields, row, converter);
        if (e != null) {
          dataList.add(e);
        }
      }
    }
    return dataList;
  }

  public <F, E> E readRow(Class<F> cls, ExcelField<F, ?>[] excelFields, Row row,
      ExcelConverter<F, E> converter)
      throws InstantiationException, IllegalAccessException, ExcelImportException {
    boolean notAllBlank = false;
    int column = 0;
    F o = cls.newInstance();
    List<CellError> rowErrors = new ArrayList<>();
    rowNum = row.getRowNum();

    for (ExcelField<F, ?> excelField : excelFields) {
      Object cellValue = excelField.getCellValue(row, column++);
      notAllBlank = notAllBlank || !excelField.isEmptyCell(cellValue);
      try {
        excelField.setProperty(o, cellValue, validator, validateGroups);
      } catch (Exception e) {
        rowErrors.add(new CellError(rowNum, column - 1, excelField.title(),
            (cellValue == null ? null : String.valueOf(cellValue)), e));
      }
    }
    if (notAllBlank) {
      if (!rowErrors.isEmpty()) {
        Exception exception = rowErrors.get(0).getException();
        throw new ExcelImportException(exception.getMessage(), rowErrors, exception);
      }
      return converter.convert(o);
    } else {
      return null;
    }
  }

  /**
   * @param validateGroups 验证 groups
   * @return ExcelImport this
   */
  public ExcelImport validateGroups(Class<?>... validateGroups) {
    this.validateGroups = validateGroups;
    return this;
  }


}

File: tools/excel/src/main/kotlin/top/bettercode/summer/tools/excel/ExcelImport.kt
package top.bettercode.summer.tools.excel

import org.dhatim.fastexcel.reader.*
import org.slf4j.LoggerFactory
import org.springframework.web.multipart.MultipartFile
import top.bettercode.summer.tools.excel.ExcelImportException.CellError
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Paths
import javax.validation.Validation
import javax.validation.groups.Default
import kotlin.streams.asSequence

/**
 * 导入Excel文件
 */
class ExcelImport private constructor(`is`: InputStream) {
    /**
     * 工作表对象
     */
    private var workbook: ReadableWorkbook = try {
        ReadableWorkbook(`is`)
    } catch (e: IOException) {
        throw ExcelException("Excel读取失败，仅支持.xlsx格式Excel文件", e)
    } catch (e: ExcelReaderException) {
        throw ExcelException("Excel读取失败，仅支持.xlsx格式Excel文件", e)
    }

    /**
     * 验证 groups
     */
    private var validateGroups = arrayOf<Class<*>>(Default::class.java)

    /**
     * 当前行号
     */
    var row = 1
        private set

    /**
     * 当前单元格号
     */
    var column = 0
        private set

    /**
     * 工作表对象
     */
    var sheet: Sheet?
        private set

    /**
     * 构造函数
     *
     * @param `is` is
     */
    init {
        try {
            sheet = workbook.firstSheet
        } catch (e: Exception) {
            throw ExcelException("Excel读取失败，文档中没有工作表", e)
        }
        setRowAndColumn(1, 0)
        log.debug("Initialize success.")
    }


    /**
     * @param row 行号，从0开始
     * @return this ExcelExport
     */
    fun setRow(row: Int): ExcelImport {
        this.row = row
        return this
    }

    /**
     * @param column 列号，从0开始
     * @return this ExcelExport
     */
    fun setColumn(column: Int): ExcelImport {
        this.column = column
        return this
    }

    /**
     * @param row    行号，从0开始
     * @param column 列号，从0开始
     * @return this ExcelExport
     */
    fun setRowAndColumn(row: Int, column: Int): ExcelImport {
        this.row = row
        this.column = column
        return this
    }

    fun sheet(sheetIndex: Int): ExcelImport {
        sheet = workbook.getSheet(sheetIndex)
            .orElseThrow { ExcelException("未找到第" + (sheetIndex + 1) + "张表") }
        setRowAndColumn(1, 0)
        return this
    }

    fun sheet(sheetName: String): ExcelImport {
        sheet = workbook.findSheet(sheetName)
            .orElseThrow { ExcelException("未找到表：$sheetName") }
        setRowAndColumn(1, 0)
        return this
    }

    /**
     * @param validateGroups 验证 groups
     * @return ExcelImport this
     */
    fun validateGroups(vararg validateGroups: Class<*>): ExcelImport {
        this.validateGroups = validateGroups.toList().toTypedArray()
        return this
    }

    /**
     * 获取导入数据列表
     *
     * @param cls         实体类型
     * @param excelFields excelFields
     * @param converter   F 转换为E
     * @param <F>         F
     * @param <E>         E
     * @return List
     * @throws IOException            IOException
     * @throws IllegalAccessException IllegalAccessException
     * @throws ExcelImportException   ExcelImportException
     * @throws InstantiationException InstantiationException
    </E></F> */
    @JvmOverloads
    fun <F, E> getData(
        excelFields: Array<ExcelField<F, *>>,
        cls: Class<F> = getEntityType(excelFields.asIterable()),
        converter: ((F) -> E)? = null
    ): List<E> {
        return getData(excelFields = excelFields.asIterable(), cls = cls, converter = converter)
    }

    /**
     * 获取导入数据列表
     *
     * @param cls         实体类型
     * @param excelFields excelFields
     * @param converter   F 转换为E
     * @param <F>         F
     * @param <E>         E
     * @return List
     * @throws IOException            IOException
     * @throws IllegalAccessException IllegalAccessException
     * @throws ExcelImportException   ExcelImportException
     * @throws InstantiationException InstantiationException
    </E></F> */
    @JvmOverloads
    fun <F, E> getData(
        excelFields: Iterable<ExcelField<F, *>>,
        cls: Class<F> = getEntityType(excelFields),
        converter: ((F) -> E)? = null
    ): List<E> {
        workbook.use {
            if (sheet == null) {
                throw RuntimeException("文档中未找到相应工作表!")
            } else {
                sheet!!.openStream().use {
                    return it.asSequence()
                        .filter { row: Row? -> row != null && row.rowNum > this.row }
                        .mapNotNull { row ->
                            readRow(
                                row = row,
                                cls = cls,
                                excelFields = excelFields,
                                converter = converter
                            )
                        }.toList()
                }
            }
        }
    }

    @JvmOverloads
    fun <F, E> readRow(
        row: Row,
        excelFields: Iterable<ExcelField<F, *>>,
        cls: Class<F> = getEntityType(excelFields),
        converter: ((F) -> E)? = null
    ): E? {
        var notAllBlank = false

        val entity = cls.getDeclaredConstructor().newInstance()

        val rowErrors: MutableList<CellError> = ArrayList()

        this.row = row.rowNum

        val validators = mutableMapOf<ExcelField<F, *>, Pair<Int, Any?>>()
        excelFields.forEachIndexed { index, excelField ->
            if (excelField.isIndexColumn) {
                return@forEachIndexed
            }
            val column = this.column + index
            val cellValue = row.getCellValue(column, excelField.isDateField)
            notAllBlank = notAllBlank || !excelField.isEmptyCell(cellValue)
            try {
                excelField.setProperty(entity, cellValue, validator, validateGroups)
            } catch (e: Exception) {
                rowErrors.add(
                    CellError(
                        row = this.row,
                        column = column,
                        excelField = excelField,
                        value = cellValue,
                        exception = e
                    )
                )
            }
            val validator = excelField.validator
            if (validator != null) {
                validators[excelField] = column to cellValue
            }
        }
        //validator
        validators.forEach { (excelField, pair) ->
            val validator = excelField.validator
            try {
                validator?.accept(entity)
            } catch (e: Exception) {
                rowErrors.add(
                    CellError(
                        row = this.row,
                        column = pair.first,
                        excelField = excelField,
                        value = pair.second,
                        exception = e
                    )
                )
            }
        }
        return if (notAllBlank) {
            if (rowErrors.isNotEmpty()) {
                val exception = rowErrors[0].exception
                throw ExcelImportException(exception.message, rowErrors, exception)
            }
            @Suppress("UNCHECKED_CAST")
            converter?.invoke(entity) ?: (entity as E)
        } else {
            null
        }
    }


    /**
     * 获取单元格值
     *
     * @param row    获取的行
     * @param column 获取单元格列号
     * @return 单元格值
     */
    fun Row.getCellValue(column: Int, isDateField: Boolean): Any? {
        return getOptionalCell(column).map {
            when (it.type) {
                CellType.STRING -> it.asString()
                CellType.NUMBER -> if (isDateField) {
                    it.asDate()
                } else {
                    it.asNumber()
                }

                CellType.BOOLEAN -> it.asBoolean()
                else -> it.value
            }
        }.orElse(null)
    }

    companion object {
        private val log = LoggerFactory.getLogger(ExcelImport::class.java)
        private val validator = Validation.buildDefaultValidatorFactory().validator

        /**
         * @param fileName 导入文件
         * @return ExcelImport
         * @throws IOException IOException
         */
        @JvmStatic
        fun of(fileName: String): ExcelImport {
            return of(Files.newInputStream(Paths.get(fileName)))
        }

        /**
         * @param file 导入文件对象
         * @return ExcelImport
         * @throws IOException IOException
         */
        @JvmStatic
        fun of(file: File): ExcelImport {
            return of(Files.newInputStream(file.toPath()))
        }

        /**
         * @param multipartFile 导入文件对象
         * @return ExcelImport
         * @throws IOException IOException
         */
        @JvmStatic
        fun of(multipartFile: MultipartFile): ExcelImport {
            return of(multipartFile.inputStream)
        }

        /**
         * @param is 输入流
         * @return ExcelImport
         * @throws IOException IOException
         */
        @JvmStatic
        fun of(`is`: InputStream): ExcelImport {
            return ExcelImport(`is`)
        }

        private fun <F> getEntityType(excelFields: Iterable<ExcelField<F, *>>): Class<F> {
            for (excelField in excelFields) {
                val entityType = excelField.entityType
                if (entityType != null) {
                    return entityType
                }
            }
            throw ExcelException("识别表单类型失败")
        }
    }
}
